### PR TITLE
(Feature) UI Improvements in Invest page

### PR DESCRIPTION
**Pull request by fvictorio**
Depends on #199.

This PR builds upon #199 and adds some improvements:

- Modify the method signature. Use `0xa6f2ae3a`, that corresponds to `buy()`
- Add QR with the address of the crowdsale
- Disable MetaMask option if web3 is not injected
- Hide Invest button when MetaMask is not selected

### CODE REVIEW

File: src/components/invest/index.js [hunk 1]
```diff
@@ -7,17 +7,23 @@ import { noMetaMaskAlert, noContractAlert, investmentDisabledAlert, investmentDi
...
       window.scrollTo(0, 0);
       if (this.tokensToInvestOnChange.bind) this.tokensToInvestOnChange = this.tokensToInvestOnChange.bind(this);
       if (this.investToTokens.bind) this.investToTokens = this.investToTokens.bind(this);
       var state = defaultState;
       state.seconds = 0;
       state.loading = true;
       state.pristineTokenInput = true;
+      state.investThrough = 'metamask'
+      state.web3Available = false;
+      state.crowdsaleAddress = ICOConfig.crowdsaleContractURL || getURLParam("addr")
```
Comment (15chrjef) [inline src/components/invest/index.js]: Extract lines 21-26 to an object in a constant or util file and use Object.assign to add it to the default state on line 20

File: src/components/invest/index.js [hunk 2]
```diff
@@ -26,16 +32,18 @@ export class Invest extends React.Component {
...
+        });
         return
       };
 
       const networkID = ICOConfig.networkID?ICOConfig.networkID:getQueryVariable("networkID");
       const contractType = this.state.contractTypes.whitelistwithcap;// getQueryVariable("contractType");
       checkNetWorkByID(web3, networkID);
       newState.contractType = contractType;
+      newState.web3Available = true;
+      newState.investThrough = 'metamask';
```
Comment (15chrjef) [inline src/components/invest/index.js]: extract string to constant

File: src/components/invest/index.js [hunk 3]
```diff
@@ -370,16 +417,27 @@ export class Invest extends React.Component {
               <div className="invest-form-label">TOKENS</div>
               {invalidTokenDescription}
             </div>
-            <a className="button button_fill" onClick={this.investToTokens}>Invest now</a>
+            <div className="invest-through-container">
+              <select value={this.state.investThrough} className="invest-through" onChange={(e) => this.setState({ investThrough: e.target.value })}>
+                <option disabled={!this.state.web3Available} value="metamask">Metamask {!this.state.web3Available ? ' (not available)' : null}</option>
```
Comment (15chrjef) [inline src/components/invest/index.js]: extract ternary to variable

**Comment by fvictorio** [comment]
Let's try to merge this as soon as possible, so that we can work on solving the MobX PR conflicts. I will create issues for all the non-critical stuff that is pending.

**Comment by fvictorio** [comment]
1. Sorry about that, it got lost during the merge conflict resolution.
2. I didn't get this. What do you mean? That hash is the signature of the `buy()` method, did that change?

**Comment by vbaranov** [comment]
2. The methodID should be: 0xa6f2ae3a

screen shot 2017-11-15 at 18 43 26

**Comment by fvictorio** [comment]
Fixed, thanks. Just curious, to what method corresponds that hash?

**Comment by vbaranov** [comment]
`0xa6f2ae3a` is for `buy` method. Do you mean wrong hash? I guess, it was just a mock hash in `invest-page-3` branch, because it was created long ago.

**Pull request merged**